### ISSUE TITLE
Clip using line stroke (not the fill)

### ISSUE THREAD

**Issue by CPatchane** [konvajs/konva#1699]
Hello, is this possible to use `clipFunc` to clip a content only on the line with the stroke width but not the fill inside?

In this [fiddle](https://jsfiddle.net/mbq25ora/), I tried to draw a line, but the clip is working on the polygon inside instead of the lines even without specifying a fill.

Here is what I have:
<img width="218" alt="Screenshot 2024-01-05 at 09 46 09" src="https://github.com/konvajs/konva/assets/10224453/2c34de07-5aea-4415-879f-cdc0a13347c6">

Here what I would expect/want:
<img width="307" alt="Screenshot 2024-01-05 at 09 52 30" src="https://github.com/konvajs/konva/assets/10224453/d0fb5c8e-3d57-42ac-942f-ce54891a0ba0">

Is it possible to achieve that using Konva? 

Thanks a lot!

**Comment by lavrton** [comment]
No, 2d canvas doesn't support stroke for clip. You have two options:

1. Draw such a line using fill, it means you will need to draw a path on the contour of a line, then fill it. Depending on the complexity of a line, it may not be a simple task.

2. Use `globalCompositeOperation` to cut the line from the rest of the shapes. You may need to cache the group, to isolate `globalCompositeOperation`: https://jsfiddle.net/y4xLf0wt/1/

**Comment by CPatchane** [comment]
Oh wow thanks for that helpful and very quick answer 🙏 

I want to do something a bit more complex, is it possible to keep the background without using a new Layer?
Here is my JSBin: https://jsfiddle.net/endjoy6g/6/

I have two issues, the groupe of lines seems to be cropped when using caching with global, don't know why:

<img width="253" alt="Screenshot 2024-01-08 at 14 55 10" src="https://github.com/konvajs/konva/assets/10224453/0624c9fa-ceb1-41f5-be53-097b365edc2a">
<img width="258" alt="Screenshot 2024-01-08 at 14 55 35" src="https://github.com/konvajs/konva/assets/10224453/4ac9c69c-b84e-4b31-8558-dd289ee27e93">

And since the caching addition, I can't drag the shape anymore event by using `draggable: true`.

The main idea here is to have a shape that help discovering a new content (green/orange) above an existing content (blue/red).

Thanks a lot 🙏

**Comment by lavrton** [comment]
You have to re-cache on dragging, to update the cached canvas. But it may drop performance on large cache.

```js
groupCut.on('dragmove', () => {
  group.cache();
});
```

https://jsfiddle.net/1o7ewksa/1/

To fix cutting, you need to increase cache area.

`groupCut.cache({ offset: 40 });`

**Comment by CPatchane** [comment]
Oh ok, very clear!
Thanks again for this very helpful response 🙏

**Comment by CPatchane** [comment]
In case someone want to have that code in React using `react-konva` ([sandbox](https://stackblitz.com/edit/af-react-sandbox-tvhat5?file=index.js)):

```jsx
const App = () => {
  const groupCutRef = useRef();
  const mainGroupRef = useRef();

  const handleDragMove = () => {
    mainGroupRef.current.cache();
  };

  useEffect(() => {
    groupCutRef.current.cache({ offset: 40 });
    mainGroupRef.current.cache();
  }, []);

  return (
    <Stage width={window.innerWidth} height={window.innerHeight}>
      <Layer>
        <Group>
          <Group>
            <Line
              points={[73, 140, 340, 23, 500, 109, 300, 170]}
              stroke="blue"
              strokeWidth={10}
              fill="#aaf"
              tension={0.8}
              closed
            />
            <Line
              points={[73, 140, 340, 23, 500, 109]}
              stroke="red"
              strokeWidth={10}
              fill="#faa"
              tension={1.2}
              scale={{ x: 0.5, y: 0.5 }}
              x={100}
              y={50}
              closed
            />
          </Group>
          <Group ref={mainGroupRef}>
            <Line
              points={[73, 140, 340, 23, 500, 109, 300, 170]}
              stroke="green"
              strokeWidth={10}
              fill="#afa"
              tension={0.8}
              closed
            />
            <Line
              points={[73, 140, 340, 23, 500, 109]}
              stroke="orange"
              strokeWidth={10}
              fill="#ffa"
              tension={1.2}
              scale={{ x: 0.5, y: 0.5 }}
              x={100}
              y={50}
              closed
            />
            <Group
              draggable
              globalCompositeOperation="destination-in"
              ref={groupCutRef}
              onDragMove={handleDragMove}
            >
              <Line
                strokeWidth={20}
                stroke="black"
                points={[
                  200, 90, 150, 110, 250, 120, 120, 120, 120, 140, 150, 80,
                ]}
              />
              <Line
                strokeWidth={20}
                stroke="red"
                points={[
                  250, 90, 200, 110, 300, 120, 170, 120, 170, 140, 200, 80,
                ]}
              />
            </Group>
          </Group>
        </Group>
      </Layer>
    </Stage>
  );
};
```

**Comment by CPatchane** [comment]
👋 @lavrton, I would have a question regarding this specific case. I guess, adding a transformer to be able to resize each line in the group `groupCut` by using an external transformer (to not be affected by the globalCompositeOperation) would be impossible here since both the transformed node and the transformer should have the same parent, right?

**Comment by lavrton** [comment]
No. You can place transformer in a very different parent.
